### ISSUE TITLE
Bandwidth/rate tuning advice

### ISSUE THREAD

**Issue by jhart-r7** [zmap/zmap#480]
I have two systems that are not identical hardware-wise but both have legit gigabit connectivity.   OS-wise they are effectively identical.   I get wildly differing results performance-wise when it comes to zmap and I attribute this to the hardware differences, both NIC and CPU.  Currently I'm trying to find a way to tune zmap such that I can maximize each host and the network capacity assigned to it, and honing in on an acceptable  `-r` or `-B` option such that zmap runs to completion.  I'll run that for a while and wait until it dies progressively further along as I decrease the rate:

```
11:07 20% (45m left); send: 362187051 543 Kp/s (543 Kp/s avg); recv: 2107142 3.14 Kp/s (3.16 Kp/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.58%
Oct 09 15:51:06.417 [FATAL] send: send rate exceeds system capabilities
```

Using consistent hardware would help here.  As might `PF_RING`.

Initially I was hoping for advice on how to tune zmap beyond just lowering the rate, but in writing this ticket I realized that maybe that is my only option.    

Does anyone have any suggestions?  Is this a bug?

Any feedback would be appreciated.

**Comment by paul-pearce** [comment]
Hmm. I'm going to include a bunch of thoughts for others as well as posterity. Specific suggestions for you towards the end.

"send rate exceeds system capabilities" is a new error and the result of the issue you reported last week about `rate` running amuck. 

The way zmap imposes rate limiting is a busy loop. How long it spins for is `delay`. At regular intervals we calculate how we'll we're doing with respect to the specified `rate` (`-r`) and update delay (if we're sending too fast, add more delay. Too slow, less delay.)

The error you're hitting occurs when delay essentially goes to zero. This means that were sending as fast as we can and can't hit `rate`. We need this because the convergence algorithm we use for `delay` is problematic, and can't tolerate delay going to zero (if it does, the rate issue you reported occurs).

The fact that this is happening _during_ a run makes me think either:

* You're operating _really_ close to system limits and eventually just get unlucky
* Some background system process is suddenly using too much CPU, and suddenly zmap isn't meeting its performance needs.

The second point above is something I was afraid might happen. I think this means we need to raise priority on fixing the convergence properties on the delay algorithm.

In terms of diagnosing why this is happening, I'd suggest looking to see what else is happening on your system around that time. Was there some cron, another scan, or system task that hit? I'd also run without `-r` and see what you stabilize at.

In terms of *solving* the problem: Try using more sender threads (`-T`). I'm 90% sure that will solve your problem. Even going to 2 will cut the workload in half and resolve this delay mess. Recall during the last issue that you pinpointed the problem as having started when we went from the default number of threads from `N-2` to 1 (where N is number of cores).

We should probably change the error msg to "send rate exceeds system capabilities at current number of threads"

**Comment by jhart-r7** [comment]
Thanks @paul-pearce.  The systems in question here are definitely operating right near the boundary of what they can handle and your guess that cron or something else is periodically spinning up and chewing that remaining % sounds like a good one.  These systems exist only for isolated zmap runs one at a time but likely have sporadic background tasks that I should axe.

Part of the problem here is that I have to use `-r`, at least in this particular case.  The systems in question have a specific packet rate that their network will not allow them to exceed without punishment, and I know that they are more than capable of, at least for some amount of time, rates much higher than what I am hardcoding because I ran zmap on these same nodes before finding https://github.com/zmap/zmap/issues/475 and they pushed 2-4x the currently hardcoded rate.  CPU was through the roof, alerts were going off and `zmap` was dropping packets like crazy, but it didn't die.

I'll try a larger `-T` after moving to beefier systems.

**Comment by paul-pearce** [comment]
> I know that they are more than capable of, at least for some amount of time ... alerts were going off and zmap was dropping packets like crazy, but it didn't die.

Ya that's the tradeoff we had to make with the prior PR. Previously zmap will just run all out to hell with the consequences (within `rate`). But that had the side effect of causing `rate` to not be respected in some corner cases. Ideally we should fix it, but until we figure out how, we now just die if things go nuts.

I'd give `-T` a try even on the less beefy systems. I expect at `-T 2` you'll be able to hit those rates again without having the system shut down.

**Comment by zakird** [comment]
We've tried to setup a slightly different approach here: https://github.com/zmap/zmap/pull/481

@jhart-r7 would you be willing to try the code in this branch and see what that looks like in your situation?

**Comment by jhart-r7** [comment]
Thanks @paul-pearce and @zakird.  I'm commencing with testing of this now.  I should have an answer in the next few hours (this only occurs at scale).

**Comment by zakird** [comment]
Sounds good. Thank you so much for your willingness to help debug this.

**Comment by jhart-r7** [comment]
I tried 6287f215eac60a60ec072ff235eafce0662e6477 and the scan was able to complete successfully from what I can tell.  One odd side effect that I can't explain is that all 12 nodes were pushing 30kp/s up until they completed ~90% of their work, and then *every* node started dropping packets.  That could definitely be something hardware or network related that is specific to this setup, but prior to this patch I hadn't seen any of these hosts drop any packets at that rate over dozens of tests -- but then again, I also never fully completed a scan (but came within just a few % several times).

**Comment by zakird** [comment]
In what sense were they dropping packets? On the send side or the receive side? Would you be willing to run the experiment again and see if the same thing happens? It's odd that this is happening, but it _could_ be a sharding issue. How many threads were you running with?

**Comment by paul-pearce** [comment]
Just to chime in here real quick. For the purpose of testing that branch Jon should be running with 1 thread. (At more than 1 thread he's unlikely to bottom out delay)

**Comment by jhart-r7** [comment]
@zakird zmap was reporting the drops:

```
 100% (38s left); send: 303126843 40.2 Kp/s (39.9 Kp/s avg); recv: 1330452 0 p/s (175 p/s avg); drops: 1.25 Kp/s (64 p/s avg); hitrate: 0.44%
```

Each shard was done without a `-T` option for threads.  The systems in questions have two CPUs.

I'll launch another test now.  After examining the systems it is very likely that some periodic system process is chewing up the remaining 20% that these systems leave for non-zmap overhead, so the same thing that had caused current master to abort mid-scan could just as likely have affected all of these hosts towards the end of the scans running 6287f21

**Comment by zakird** [comment]
Those drops are on the receiving side (i.e., in `libpcap`) and should be unaffected by the change here. I suspect something else is awry.

**Comment by jhart-r7** [comment]
The second test completed successfully, but the same drop characteristics occurred again -- as each system reached somewhere around 90% completion of its given shard, it started dropping packets up until the end.  I definitely agree that the hardware/network resources for these systems could use some improvement, but I'm also somewhat suspicious that this same odd behavior happened twice in a row, at basically the same point, on all nodes.

I'll do some more testing by throttling these systems back a bit or giving them more resources.

**Comment by jhart-r7** [comment]
Just a quick update -- I haven't had time to continue the aforementioned testing.  I'm inclined to say 6287f21 is good to go because in using this same version on unconstrained systems I see what I expect -- fast rates, no drops.

**Comment by jhart-r7** [comment]
I've found an issue while testing 6287f21, though I suspect that this problem exists in current master too.

When using (at least) the `-B` option to constrain the rate at which zmap sends, it appears to not respect this option and sends at a much faster rate.

On two mostly identical systems, I've run zmap with:

```
zmap -i <interface> -s 123 -S <27 IPs> -v1 -b /home/scanner/blacklist.conf -B 450M --output-filter='success = 1 && repeat = 0 && sport = 123' -p 123 --seed 1508346998 --shards 1 --shard 0 -M udp -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl,data --probe-args=file:/home/scanner/ntp_123_monlist.pkt
```

Fairly quickly it stablizes at ~484Kp/s:

```
 0:09 0% (2h06m left); send: 4358579 484 Kp/s (482 Kp/s avg); recv: 1423 161 p/s (157 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.03%
```

My intention was to limit the scan to 450Mb/s, however in reality it pushes nearly double that according to `nload` and an RRD-based monitoring solution of the switch that handles this traffic.

Some quick math shows that with the 192-byte payload we are sending, if zmap thinks it should send at 484Kp/s, in reality it is sending at a rate of ~743Mb/s (8 * 192 * 484000).

**Comment by jhart-r7** [comment]
I'm going to go ahead and close this.    With the exception of what I've recently filed in #499, most of our scans have been running successfully when we pick a reasonable rate.  We've been running 91b334091d9207e4a53884c67a6f10ac696866cf for some time now, and in the few cases where we ran into scan rate issues that overwhelmed the underlying OS/hardware, it seems like there were non-zmap things causing issues.  

If this crops up again, I'll re-open with more current details.  Thanks!
